src/Simplify_Call.cpp

```diff
@@ -170,53 +171,49 @@ Expr Simplify::visit(const Call *op, ExprInfo *bounds) {
 
         const Type t = op->type;
 
-        uint64_t ub = 0;
-        int64_t sb = 0;
-        bool b_is_const_uint = const_uint(b, &ub);
-        bool b_is_const_int = const_int(b, &sb);
-        if (b_is_const_uint || b_is_const_int) {
-            if (b_is_const_int) {
-                ub = std::abs(sb);
-            }
+        // We might swap from a right to left shift or the reverse.
+        std::string result_op = op->name;
 
-            // Determine which direction to shift.
-            const bool b_is_pos = b_is_const_uint || (b_is_const_int && sb >= 0);
-            const bool b_is_neg = b_is_const_int && sb < 0;
-            const bool shift_left = ((op->is_intrinsic(Call::shift_left) && b_is_pos) ||
-                                     (op->is_intrinsic(Call::shift_right) && b_is_neg));
-            const bool shift_right = ((op->is_intrinsic(Call::shift_right) && b_is_pos) ||
-                                      (op->is_intrinsic(Call::shift_left) && b_is_neg));
+        // If we know the sign of this shift, change it to an unsigned shift.
+        if (b_info.min_defined && b_info.min >= 0) {
+            b = cast(b.type().with_code(halide_type_uint), b);
+        } else if (b_info.max_defined && b_info.max <= 0) {
+            result_op = Call::get_intrinsic_name(op->is_intrinsic(Call::shift_right) ? Call::shift_left : Call::shift_right);
+            b = mutate(cast(b.type().with_code(halide_type_uint), -b), nullptr);
+        }
 
+        // If the shift is by a constant, it should now be unsigned.
+        uint64_t ub = 0;
+        if (const_uint(b, &ub)) {
             // LLVM shl and shr instructions produce poison for
             // shifts >= typesize, so we will follow suit in our simplifier.
             if (ub >= (uint64_t)(t.bits())) {
                 return make_signed_integer_overflow(t);
             }
             if (a.type().is_uint() || ub < ((uint64_t)t.bits() - 1)) {
                 b = make_const(t, ((int64_t)1LL) << ub);
-                if (shift_left) {
+                if (result_op == Call::get_intrinsic_name(Call::shift_left)) {
                     return mutate(Mul::make(a, b), bounds);
-                } else if (shift_right) {
+                } else {
                     return mutate(Div::make(a, b), bounds);
                 }
             } else {
-                // For signed types, (1 << ub) will overflow into the sign bit while
-                // (-32768 >> ub) propagates the sign bit, making decomposition
+                // For signed types, (1 << (t.bits() - 1)) will overflow into the sign bit while
+                // (-32768 >> (t.bits() - 1)) propagates the sign bit, making decomposition
                 // into mul or div problematic, so just special-case them here.
-                if (shift_left) {
+                if (result_op == Call::get_intrinsic_name(Call::shift_left)) {
```

Comment:
`op->is_intrinsic(Call::shift_left)` seems more clear to me, here and in the check above

Reply:
We can't do that though, because we might have swapped from left to right shifts above?

Reply:
The thing that would be best is if we could just use Call::IntrinscOp instead of std::string, but that's a bit messy to do.

Reply:
Ah sorry, my bad, makes sense!